modules/http_push.c

```diff
@@ -61,30 +64,66 @@ static int nw_http_push_start_acquire_data(struct nodewatcher_module *module,
           /* Default. */
           timeout = 5;
         }
+
+        const char* data_string = json_object_to_json_string(data);
+
         curl_easy_setopt(curl, CURLOPT_TIMEOUT, timeout);
         curl_easy_setopt(curl, CURLOPT_FAILONERROR, 1);
         curl_easy_setopt(curl, CURLOPT_URL, url);
         curl_easy_setopt(curl, CURLOPT_POST, 1);
-        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, json_object_to_json_string(data));
+        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data_string);
+
+        const char *auth_type = nw_uci_get_string(uci, "nodewatcher.@agent[0].push_authentication_method");
+
+        if (strcasecmp(auth_type, "hmac") == 0) {
+          const char *hmac_key = nw_uci_get_string(uci, "nodewatcher.@agent[0].hmac_key");
+
+          if (hmac_key) {
+            unsigned char hmac_out[SHA256_HASH_SIZE];
+            hmac_sha256((unsigned char*) hmac_key, strlen(hmac_key), (unsigned char*) data_string, strlen(data_string), hmac_out);
+
+            char signature[SHA256_HASH_SIZE + 1] = {0};
+
+            if (nw_base64_encode(hmac_out, SHA256_HASH_SIZE, signature, SHA256_HASH_SIZE) == 0) {
+              char signature_header[26+sizeof(signature)] = {0};
+              strcat(strcpy(signature_header, "X-Nodewatcher-Signature: "), signature);
```

Comment:
Odd at least. Can you use snprintf?
Or, how about just initializing to "X-Nodewatcher-Signature: " instead of {0}?